<a href="https://colab.research.google.com/github/Rimuy-Amaya/Python_20250503/blob/main/lesson1/test2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
print("hi hi")

hi hi


In [ ]:
price = int (input("輸入顧客購買金額:"))
if price >= 100000:
    price = price * 0.8
elif price >= 50000:
    price = price * 0.85
elif price >= 30000:
    price = price * 0.9
elif price >= 10000:
    price = price * 0.95
print("實付金額是:",price)

輸入顧客購買金額:45623555
實付金額是: 36498844.0


In [ ]:
price = int(input("輸入顧客購買金額:"))

if price >= 100000:
    price = price * 0.8
elif price >= 50000:
    price = price * 0.85
elif price >= 30000:
    price = price * 0.9
elif price >= 10000:
    price = price * 0.95

print("實付金額是:", price)

輸入顧客購買金額:453
實付金額是: 453


In [ ]:
a=14
if a>=12 and a<18 :
	print("ok-1")
else:
	print("cancel- 1")

if 12<=a<18:
	print("ok-2")
else:
	print("cancel- 2")

if not a>18:
	print("ok-3")
else:
	print("cancel- 3")

ok-1
ok-2
ok-3


In [ ]:

height = float ( input("請輸入身高，單位為公分:"))
weight = float ( input("請輸入體重，單位為公分:"))
BMI = weight / ( (height/100)**2 )

if BMI < 18.5:
    print(f"您的BMI是{BMI}\n「您的體重太輕」")
elif BMI < 25:
    print(f"您的BMI是{BMI}\n「您的體重正常」")
elif BMI < 30:
    print(f"您的BMI是{BMI}\n「您的體重過重」")
else:
    print(f"您的BMI是{BMI}\n「您的體重肥胖」")

請輸入身高，單位為公分:163
請輸入體重，單位為公分:62
您的BMI是23.335466144755166
「您的體重正常」


In [ ]:
score_forall=0

for score in range(0,5):
    score = int( input("請輸入學生的成績:"))
    score_forall += score
score_average = score_forall / 5

print(f"全班總成績為:{score_forall}分，平均為{score_average}分")

請輸入學生的成績:55
請輸入學生的成績:55
請輸入學生的成績:55
請輸入學生的成績:55
請輸入學生的成績:55
全班總成績為:275分，平均為55.0分


In [ ]:
sum = 0
students = int( input("請輸入學生總人數:") )
for i in range(students):
    studentScore = int(input('請輸入第'+ str(i+1) +'位學生的成績:'))
    sum += studentScore

print("全班總成績為:" + str(sum) + "分,平均分數為" + str(sum/students) + "分")

請輸入學生總人數:7
請輸入第1位學生的成績:55
請輸入第2位學生的成績:32
請輸入第3位學生的成績:12
請輸入第4位學生的成績:80
請輸入第5位學生的成績:100
請輸入第6位學生的成績:15
請輸入第7位學生的成績:55
全班總成績為:349分,平均分數為49.857142857142854分


In [ ]:
print("兩個參數")
for i in range(4,8):
    print(i)
print("離開後i為",i)

兩個參數
4
5
6
7
離開後i為 7


In [ ]:
print("三個參數")
for i in range(4,0,-1):
    print(i)
print("離開後i為",i)

三個參數
4
3
2
1
離開後i為 1


In [ ]:
print("三個參數")
for i in range(4,8,2):
    print(i)
print("離開後i為",i)

三個參數
4
6
離開後i為 6


In [ ]:
for x in range(0,8,2) :
    print(x)

0
2
4
6


In [ ]:
total = 0
for i in range(1,101):
    total += i

print(f"1加到100的加總為{total}")

1加到100的加總為5050


In [ ]:
end = int(input("1加到多少的數值:"))
total = 0
for i in range(1,end+1):
    total += i

print(f"1加到{end}的加總為{total}")

1加到多少的數值:5000
1加到5000的加總為12502500


In [ ]:
start = int(input("start的數值:"))
end = int(input("結束的數值:"))
total = 0
for i in range(start,end+1):
    total += i

print(f"{start}加到{end}的加總為{total}")

start的數值:10
結束的數值:-5
10加到-5的加總為0


In [ ]:
#1~100,偶數的加總
#1~100,奇數的加總
evenTotal = 0
oddTotal = 0

for i in range(1,101):
    if i % 2 == 0:
        evenTotal += i
    else:
        oddTotal += i

print(f"1加到100偶數的加總為{evenTotal}")
print(f"1加到100奇數的加總為{oddTotal}")

1加到100偶數的加總為2550
1加到100奇數的加總為2500


In [ ]:
for i in range(1,10):
    for j in range(1,10):
        print(f"{i:<2d}*{j:2d} ={i*j:2d}",end='  ')
    print()

1 * 1 = 1  1 * 2 = 2  1 * 3 = 3  1 * 4 = 4  1 * 5 = 5  1 * 6 = 6  1 * 7 = 7  1 * 8 = 8  1 * 9 = 9  
2 * 1 = 2  2 * 2 = 4  2 * 3 = 6  2 * 4 = 8  2 * 5 =10  2 * 6 =12  2 * 7 =14  2 * 8 =16  2 * 9 =18  
3 * 1 = 3  3 * 2 = 6  3 * 3 = 9  3 * 4 =12  3 * 5 =15  3 * 6 =18  3 * 7 =21  3 * 8 =24  3 * 9 =27  
4 * 1 = 4  4 * 2 = 8  4 * 3 =12  4 * 4 =16  4 * 5 =20  4 * 6 =24  4 * 7 =28  4 * 8 =32  4 * 9 =36  
5 * 1 = 5  5 * 2 =10  5 * 3 =15  5 * 4 =20  5 * 5 =25  5 * 6 =30  5 * 7 =35  5 * 8 =40  5 * 9 =45  
6 * 1 = 6  6 * 2 =12  6 * 3 =18  6 * 4 =24  6 * 5 =30  6 * 6 =36  6 * 7 =42  6 * 8 =48  6 * 9 =54  
7 * 1 = 7  7 * 2 =14  7 * 3 =21  7 * 4 =28  7 * 5 =35  7 * 6 =42  7 * 7 =49  7 * 8 =56  7 * 9 =63  
8 * 1 = 8  8 * 2 =16  8 * 3 =24  8 * 4 =32  8 * 5 =40  8 * 6 =48  8 * 7 =56  8 * 8 =64  8 * 9 =72  
9 * 1 = 9  9 * 2 =18  9 * 3 =27  9 * 4 =36  9 * 5 =45  9 * 6 =54  9 * 7 =63  9 * 8 =72  9 * 9 =81  


In [ ]:
for i in range(1,10):
    for j in range(1,10):
        print(f"{i:<2d}*{j:2d}={i*j:2d}",end = " ")
    print()

1 * 1= 1 1 * 2= 2 1 * 3= 3 1 * 4= 4 1 * 5= 5 1 * 6= 6 1 * 7= 7 1 * 8= 8 1 * 9= 9 
2 * 1= 2 2 * 2= 4 2 * 3= 6 2 * 4= 8 2 * 5=10 2 * 6=12 2 * 7=14 2 * 8=16 2 * 9=18 
3 * 1= 3 3 * 2= 6 3 * 3= 9 3 * 4=12 3 * 5=15 3 * 6=18 3 * 7=21 3 * 8=24 3 * 9=27 
4 * 1= 4 4 * 2= 8 4 * 3=12 4 * 4=16 4 * 5=20 4 * 6=24 4 * 7=28 4 * 8=32 4 * 9=36 
5 * 1= 5 5 * 2=10 5 * 3=15 5 * 4=20 5 * 5=25 5 * 6=30 5 * 7=35 5 * 8=40 5 * 9=45 
6 * 1= 6 6 * 2=12 6 * 3=18 6 * 4=24 6 * 5=30 6 * 6=36 6 * 7=42 6 * 8=48 6 * 9=54 
7 * 1= 7 7 * 2=14 7 * 3=21 7 * 4=28 7 * 5=35 7 * 6=42 7 * 7=49 7 * 8=56 7 * 9=63 
8 * 1= 8 8 * 2=16 8 * 3=24 8 * 4=32 8 * 5=40 8 * 6=48 8 * 7=56 8 * 8=64 8 * 9=72 
9 * 1= 9 9 * 2=18 9 * 3=27 9 * 4=36 9 * 5=45 9 * 6=54 9 * 7=63 9 * 8=72 9 * 9=81 


In [ ]:
for i in range(1,6):
    for _ in range(i):
        print("#",end='')
    print()

#
##
###
####
#####


In [ ]:
#求所有因數
num = int(input("請輸入整數，求所有因數:"))
print(num,"的因數:")
for i in range(1,num+1):
  if num % i == 0:
    print(i)

請輸入整數，求所有因數:456
456 的因數:
1
2
3
4
6
8
12
19
24
38
57
76
114
152
228
456


In [ ]:
#求公因數
print("請輸入2數,求2數公因數:")
n = int(input('請輸入第1個數:'))
m = int(input('請輸入第2個數:'))
if n > m:
  num = m
else:
  num = n

print(n,'和',m, '的公因數是:')
for i in range(1,num+1):
  if n % i == 0 and m % i == 0:
    print(i)

請輸入2數,求2數公因數:
請輸入第1個數:55
請輸入第2個數:5
55 和 5 的公因數是:
1
5


In [ ]:
#求質數
end = int(input("請輸入end值:"))
print(f"1到{end}的質數是:")
for num in range(2,end+1):
    is_prime = True
    for i in range(2,num):
        if num % i == 0:
            is_prime = False

    if is_prime:
        print(f"{num}",end=" ")

請輸入end值:4531
1到4531的質數是:
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 53 59 61 67 71 73 79 83 89 97 101 103 107 109 113 127 131 137 139 149 151 157 163 167 173 179 181 191 193 197 199 211 223 227 229 233 239 241 251 257 263 269 271 277 281 283 293 307 311 313 317 331 337 347 349 353 359 367 373 379 383 389 397 401 409 419 421 431 433 439 443 449 457 461 463 467 479 487 491 499 503 509 521 523 541 547 557 563 569 571 577 587 593 599 601 607 613 617 619 631 641 643 647 653 659 661 673 677 683 691 701 709 719 727 733 739 743 751 757 761 769 773 787 797 809 811 821 823 827 829 839 853 857 859 863 877 881 883 887 907 911 919 929 937 941 947 953 967 971 977 983 991 997 1009 1013 1019 1021 1031 1033 1039 1049 1051 1061 1063 1069 1087 1091 1093 1097 1103 1109 1117 1123 1129 1151 1153 1163 1171 1181 1187 1193 1201 1213 1217 1223 1229 1231 1237 1249 1259 1277 1279 1283 1289 1291 1297 1301 1303 1307 1319 1321 1327 1361 1367 1373 1381 1399 1409 1423 1427 1429 1433 1439 1447 1451 1453 1459 1471 1481 14

In [ ]:
#檢查是否cheeses為空list
cheeses = []
for cheese in cheeses:
    print('我喜歡的cheese有', cheese)
    break
else: # no break means no cheese
    print('沒有任何cheese')

沒有任何cheese


In [ ]:
a=6
while a > 0:
	print(a)
	a -= 1
print("離開後a為",a)

6
5
4
3
2
1
離開後a為 0


In [ ]:
a = 15
while a > 0:
	print(a)
	a -= 2

15
13
11
9
7
5
3
1


In [ ]:
sum = 0
i = 2
while(i <= 10):
    sum += i
    print("第",i/2,"次迴圈的i =",i,"總和為", sum);
    i += 2

第 1.0 次迴圈的i = 2 總和為 2
第 2.0 次迴圈的i = 4 總和為 6
第 3.0 次迴圈的i = 6 總和為 12
第 4.0 次迴圈的i = 8 總和為 20
第 5.0 次迴圈的i = 10 總和為 30


# 變數命名
## 變數命名：inputNum 和 deposit
- 在程式設計中，給變數取名字不僅僅是為了讓程式碼能執行，更重要的是為了讓程式碼容易閱讀、理解和維護，無論是對你自己還是對其他開發者而言。

**deposit (總存款)**
- 命名慣例： deposit 這個名字是非常常見且直觀的，尤其是在處理金錢或儲蓄相關的程式中。它清晰地表示這個變數儲存的是「存款總額」。

背後思考：
- 語義清晰： deposit 直接對應現實生活中的「存款」，讓人一眼就能明白它的作用。
- 狀態變數： 它代表了小明「目前的」總存款狀態，這個狀態會隨著時間（每月存錢）而改變。它是一個累積的量。
- 避免歧義： 如果把它叫做 total 或 sum，雖然也能表達總和，但在這個情境下，deposit 更具體，更能傳達是「錢」的總和。

**inputNum (當月輸入的金額)**
- 命名慣例： inputNum 也是一個常見且具描述性的命名方式。它結合了「輸入 (input)」和「數字 (Num)」兩個概念。

背後思考：
- 來源清晰： inputNum 清楚地表明這個值是來自使用者的「輸入」。
- 臨時性 / 局部性： 這個變數通常只在迴圈的單次迭代中有效，它暫時儲存了當前月份使用者輸入的金額。它是一個「當次」的量，而不是累積的量。
- 與總和區分： 將 inputNum 和 deposit 區分開來，可以讓程式碼的邏輯更清晰：
我從使用者那裡「取得」了 inputNum（當月金額）。
然後我把 inputNum「加到」 deposit（總存款）裡面。 如果直接 deposit += int(input(...))，雖然可以，但少了一個中間變數來明確表示「當月輸入的這筆錢」。
總結

deposit 和 inputNum 這些變數名稱的設計，並不是絕對的「慣例」（例如 i, j 那樣的固定用法），而是遵循了程式設計中非常重要的**「描述性命名 (Descriptive Naming)」**原則。

目的： 讓程式碼自我解釋，提高可讀性。
思考方式： 變數名稱應該能夠清楚地傳達它儲存了什麼資料，以及它在程式中扮演什麼角色。
一個好的變數名稱就像一個小小的註解，不需要看額外的說明，就能讓人理解它的作用。在你的程式碼中，deposit 代表累計的總存款，而 inputNum 代表單次輸入的月份存款，兩者各司其職，命名得非常恰當。

In [ ]:
#============================================================================
# Name        : while2.py
# 小明想要存錢買一輛機車,機車每輛30000元，他將每月存的錢輸入，當存款足夠買機車時，就顯示提示訊息告知。
#============================================================================
deposit = 0
num = 0
while(deposit < 30000):
    num += 1
    inputNum = int(input('請輸入第'+str(num)+"個月份的存款:")) # 這行就是輸入存款金額的地方
    deposit += inputNum
print("恭喜!已經存夠了，存了",num,"個月的總存款為:",deposit,"元。")

請輸入第1個月份的存款:456
請輸入第2個月份的存款:4532
請輸入第3個月份的存款:4536
請輸入第4個月份的存款:5555
請輸入第5個月份的存款:5555
請輸入第6個月份的存款:5555
請輸入第7個月份的存款:5555
恭喜!已經存夠了，存了 7 個月的總存款為: 31744 元。


In [ ]:
#使用傳統的方式,讀取list內的每一個元素
rabbits = ['Flopsy', 'Mopsy', 'Cottontail', 'Peter']
current = 0
while current < len(rabbits):
    print(rabbits[current])
    current += 1

Flopsy
Mopsy
Cottontail
Peter


# for_in迴圈
為什麼這比 while 迴圈更方便？
- 自動管理索引： 你不需要手動創建和更新索引變數（像 current 那樣）。Python 替你處理了元素一個一個取出的細節。
- 程式碼更簡潔： 少了初始化索引、檢查長度、遞增索引的程式碼，使得邏輯更清晰，更不容易出錯。
- 更具語義： for rabbit in rabbits 直接讀起來就是「對於 rabbits 列表中的每一個 rabbit 而言...」，非常符合人類的思考習慣。

這就是 Python 這種高階語言的設計理念之一：**讓程式碼更接近自然語言，更注重「做什麼」而不是「怎麼做」。**

In [ ]:
#使用更簡潔方式(for..in)
rabbits = ['Flopsy', 'Mopsy', 'Cottontail', 'Peter']
for rabbit in rabbits:
    print(rabbit)

Flopsy
Mopsy
Cottontail
Peter


In [ ]:
# 字串每次取出一個字元
word = 'cat'
for letter in word:
    print(letter)

c
a
t


In [ ]:
accusation = {'room': 'ballroom', 'weapon': 'lead pipe', 'person': 'Col. Mustard'}

for card in accusation: # 或者使用 accusation.keys():
    print(card)

room
weapon
person


In [ ]:
#使用values()方法取出元素的值
for value in accusation.values():
    print(value)

ballroom
lead pipe
Col. Mustard


In [ ]:
accusation = {'room': 'ballroom', 'weapon': 'lead pipe', 'person': 'Col. Mustard'}

#使用items()方法,取出包含key和value的tuple,
for item in accusation.items():
    print(item)

('room', 'ballroom')
('weapon', 'lead pipe')
('person', 'Col. Mustard')


In [ ]:
accusation = {'room': 'ballroom', 'weapon': 'lead pipe', 'person': 'Col. Mustard'}

#使用拆解法直接同時取出key和value
for card, contents in accusation.items():
    print('Card', card, 'has the contents', contents )

Card room has the contents ballroom
Card weapon has the contents lead pipe
Card person has the contents Col. Mustard


# break 的使用時機：「找到目標 / 任務完成 / 不再需要繼續」
你可以把 break 想像成一個緊急停止按鈕。當你按下它，整個迴圈就立即停止了，無論迴圈原本設計要執行多少次。

In [ ]:
i = ['a', 'b', 'c', 'd']
for j in i:
    if j == 'c':
        break
    print(j)

a
b


# continue 的使用時機：「跳過不符條件的 / 排除無效的 / 繼續處理其他」
- 你可以把 continue 想像成一個**「跳過這一次」的按鈕**。當你按下它，迴圈會跳過當前這次迭代中剩餘的程式碼，直接開始下一次迭代。所以，它不是「停止整個迴圈」，也不是「繼續執行當前迴圈的所有步驟」，而是「跳過這次的後半段，直接去進行下一次」。

- continue 這個詞在中文裡有「繼續」的意思，所以會讓人直覺以為是「繼續執行」，但它在程式裡的意思其實有點不一樣。
- 這就像你看電影，看到某個片段你不想看，就按了「跳過」(skip)。電影就跳到下一個片段，而不是整個電影停止了。continue 就是迴圈中的「跳過這一次」按鈕。

In [ ]:
i = ['a', 'b', 'c', 'd']
for j in i:
    if j == 'c':
        continue
    print(j)

a
b
d


# 總結心法：
- break： 「夠了！我不需要再做了，結束整個任務！」
- continue： 「嗯...這個有點麻煩/不對勁，跳過它，但其他的還是要繼續做下去！」

In [ ]:
days = ['Monday', 'Tuesday', 'Wednesday']
fruits = ['banana', 'orange', 'peach']
drinks = ['coffee', 'tea', 'beer']
desserts = ['tiramisu', 'ice cream', 'pie', 'pudding']
for day, fruit, drink, dessert in zip(days, fruits, drinks, desserts):
    print(day, ": drink", drink, "- eat", fruit, "- enjoy", dessert)


Monday : drink coffee - eat banana - enjoy tiramisu
Tuesday : drink tea - eat orange - enjoy ice cream
Wednesday : drink beer - eat peach - enjoy pie


In [ ]:
# 造樣造句的元素列表
persons = ["小明", "小華", "老師", "小麗"]
times = ["早上", "下午", "晚上", "中午"]
locations = ["公園", "圖書館", "學校", "家裡"]
objects = ["一本書", "一個足球", "一台筆電", "一隻鉛筆"]
actions = ["閱讀", "踢球", "寫作業", "畫畫"]

print("--- 造樣造句生成器 ---")

# 使用 zip() 同步迭代所有列表
# 注意：zip 會以最短的列表為準，所以如果列表長度不一，有些組合會被忽略。
for person, time, location, obj, action in zip(persons, times, locations, objects, actions):
    # 組合句子
    sentence = f"{person}{time}在{location}拿著{obj}{action}。"
    print(sentence)

print("\n--- 造句結束 ---")

--- 造樣造句生成器 ---
小明早上在公園拿著一本書閱讀。
小華下午在圖書館拿著一個足球踢球。
老師晚上在學校拿著一台筆電寫作業。
小麗中午在家裡拿著一隻鉛筆畫畫。

--- 造句結束 ---


# 隨機造句
你提出了一個很棒的想法！要讓造句隨機，而不是簡單的位置對應，你需要引入「隨機性」的概念。這在 Python 中可以使用 random 模組來實現。

主要思路是：每次造句時，從每個類別的列表中隨機挑選一個元素，而不是按照順序來。

讓造句隨機化
要實現這個功能，我們需要使用 Python 內建的 random 模組。具體來說，我們會用到 **random.choice() 函式**，它能從一個序列（如列表）中隨機選取一個元素。

In [ ]:
import random # 載入 random 模組，這樣我們才能使用它的隨機功能

# 造樣造句的元素列表
persons = ["小明", "小華", "老師", "小麗", "工程師", "阿嬤"]
times = ["早上", "下午", "晚上", "中午", "清晨", "深夜"]
locations = ["公園", "圖書館", "學校", "家裡", "咖啡廳", "實驗室"]
objects = ["一本書", "一個足球", "一台筆電", "一隻鉛筆", "一杯咖啡", "一個工具"]
actions = ["閱讀", "踢球", "寫作業", "畫畫", "聊天", "研究"]

print("--- 隨機造句生成器 ---")

num_sentences = 5 # 決定要生成多少個隨機句子

for _ in range(num_sentences): # 迴圈跑 num_sentences 次來生成句子
    # 從每個列表中隨機選擇一個元素
    random_person = random.choice(persons)
    random_time = random.choice(times)
    random_location = random.choice(locations)
    random_object = random.choice(objects)
    random_action = random.choice(actions)

    # 組合句子
    sentence = f"{random_person}{random_time}在{random_location}拿著{random_object}{random_action}。"
    print(sentence)

print("\n--- 造句結束 ---")

--- 隨機造句生成器 ---
小明早上在公園拿著一台筆電研究。
小麗早上在學校拿著一隻鉛筆閱讀。
老師中午在圖書館拿著一個足球閱讀。
小明清晨在學校拿著一個足球聊天。
小華中午在圖書館拿著一個工具寫作業。

--- 造句結束 ---


## 應用情境：zip()函數在以下情況非常有用：

- 資料配對： 當你有多個相關的序列（例如，學生的姓名列表和他們的成績列表），你想將它們配對起來一起處理時。
- 同步迭代： 在 for 迴圈中，當你需要同時遍歷多個序列的對應元素時（就像你之前看到的 for day, fruit, drink, dessert in zip(...)）。
- 創建字典： zip() 常用於將兩個列表（一個作為鍵，一個作為值）快速組合成一個字典。

In [ ]:
#使用zip()組合每個串列內元素成為tuple
english = 'Monday', 'Tuesday', 'Wednesday'
french = 'Lundi', 'Mardi', 'Mercredi'
list( zip(english, french))

[('Monday', 'Lundi'), ('Tuesday', 'Mardi'), ('Wednesday', 'Mercredi')]

Homework:
#問題various_loop1.py
以for迴圈計算1到100的和

顯示=================
1+2+3+~+100的總合是5050

In [21]:
num = 0

for i in range(1,101):
    num += i
print("顯示=================")
print(f"1+2+3+~+100的總合是{num}")

顯示=================
1+2+3+~+100的總合是5050


# 變數命名的思考原則
- 描述性 (Descriptive)： 這個變數是做什麼用的？它儲存了什麼數據？名稱應該能傳達這些信息。
- 清晰性 (Clarity)： 避免使用會引起歧義的名稱。
- 簡潔性 (Conciseness)： 在清晰的前提下，可以適當簡潔。但不要為了簡潔而犧牲了描述性。
- 一致性 (Consistency)： 在同一個專案或程式中，盡量保持命名風格的一致性。
避免縮寫，除非很常見： num 雖然是 "number" 的縮寫，但因為其模糊性，所以不太推薦。但像 idx (index) 或 avg (average) 這些廣為人知的縮寫就沒問題。

In [22]:
#gemini建議變數命名為total_sum
total_sum = 0
for i in range(1, 101):
    total_sum += i
print(f"1+2+3+~+100的總合是{total_sum}")

1+2+3+~+100的總合是5050


In [25]:
sum = 0
for i in range(100):
    i += 1
    if(i != 100):
        print(i,"+", end='')
        sum += i
        continue
    print(i," = ", end='')
    sum += i

print(sum)

1 +2 +3 +4 +5 +6 +7 +8 +9 +10 +11 +12 +13 +14 +15 +16 +17 +18 +19 +20 +21 +22 +23 +24 +25 +26 +27 +28 +29 +30 +31 +32 +33 +34 +35 +36 +37 +38 +39 +40 +41 +42 +43 +44 +45 +46 +47 +48 +49 +50 +51 +52 +53 +54 +55 +56 +57 +58 +59 +60 +61 +62 +63 +64 +65 +66 +67 +68 +69 +70 +71 +72 +73 +74 +75 +76 +77 +78 +79 +80 +81 +82 +83 +84 +85 +86 +87 +88 +89 +90 +91 +92 +93 +94 +95 +96 +97 +98 +99 +100  = 5050


# 使用while迴圈

In [27]:
# 初始化變數
total_sum = 0  # 用來儲存總和
current_num = 1 # 用來追蹤當前的數字，從 1 開始

print("顯示=================") # 顯示分隔線

# while 迴圈：當 current_num 小於等於 100 時，持續執行
while current_num <= 100:
    if current_num != 100:
        # 如果 current_num 不是 100 (即 1 到 99 的數字)
        print(current_num, "+", sep="", end="") # 印出數字和 "+"，不換行，且數字和 "+" 之間沒有空格
        total_sum += current_num # 將數字加到總和裡
    else:
        # 如果 current_num 是 100 (最後一個數字)
        print(current_num, " = ", sep="", end="") # 印出 100 和 " = "，不換行，且數字和 "=" 之間沒有空格
        total_sum += current_num # 將 100 加到總和裡

    current_num += 1 # 讓 current_num 增加 1，準備處理下一個數字

print(total_sum) # 迴圈結束後，印出最終的總和

顯示=================
1+2+3+4+5+6+7+8+9+10+11+12+13+14+15+16+17+18+19+20+21+22+23+24+25+26+27+28+29+30+31+32+33+34+35+36+37+38+39+40+41+42+43+44+45+46+47+48+49+50+51+52+53+54+55+56+57+58+59+60+61+62+63+64+65+66+67+68+69+70+71+72+73+74+75+76+77+78+79+80+81+82+83+84+85+86+87+88+89+90+91+92+93+94+95+96+97+98+99+100 = 5050


#問題 nestedLoop1.py
試寫出下列數字排列的程式
顯示=================================

55555
4444
333
22
1

## gemini批改修正
print(str(i) * i)

這就是核心的修正！
str(i)：首先，將數字 i 轉換成字串。例如，當 i 是 5 時，str(i) 就會變成 "5"。
* i：然後，對這個字串使用乘號 *，後面跟著數字 i。這在 Python 中表示將該字串重複 i 次。
例如：當 i 是 5 時，str(5) * 5 就是 "5" * 5，結果是 "55555"。
當 i 是 4 時，str(4) * 4 就是 "4" * 4，結果是 "4444"。

In [42]:
print("顯示=================================")

# 外層迴圈：控制每一行要印出的數字
# range(5, 0, -1) 會產生 5, 4, 3, 2, 1
for i in range(5, 0, -1):
    # 內層邏輯：控制要印出幾個數字 'i'
    # 你想印出 'i' 這個數字 i 次
    # 例如：當 i 是 5，你想印出 5 個 '5' (即 "55555")
    # 將數字 i 轉換為字串 str(i)，然後用 * 運算子重複 i 次
    print(str(i) * i)

顯示=================================
55555
4444
333
22
1


In [51]:
M = int( input("請輸入第一個數字:") )
N = int( input("請輸入第二個數字:") )
ori = 0

print("顯示==========================")
print(f"輸入M:{M}")
print(f"輸入N:{N}")
for i in range(1,N+1):
    if i < N:
        ori += i
        print( f"{M}*{i}",end="+")
    elif i == N:
        print( f"{M}*{i}",end="=")
print(M*N)

請輸入第一個數字:1 
請輸入第二個數字:10
顯示==========================
輸入M:1
輸入N:10
1*1+1*2+1*3+1*4+1*5+1*6+1*7+1*8+1*9+1*10=10


## 你的程式碼哪裡出了問題？
問題的根源在於對「總和」的計算方式理解錯誤，以及最終印出的值不對。

- 錯誤點一：總和計算錯誤 (ori += i)
你的意圖： 你想計算 M*1 + M*2 + ... + M*N 的總和。
你實際做的： ori += i 這一行，每次只把 i 的值加到 ori 裡面。
這實際上是計算了 1 + 2 + 3 + ... + (N-1) 的和（因為 elif i == N 那邊沒有加到 ori 裡面）。
正確做法： 每次迴圈，應該是把 M * i 這個「項」加到總和裡。

- 錯誤點二：最終結果輸出錯誤 (print(M*N))
你的意圖： 你想印出所有項加起來的最終總和。
你實際做的： print(M*N) 這行，只會印出最後一項 M 乘以 N 的結果，而不是前面所有項加起來的結果。
正確做法： 應該印出你用來累積總和的變數（例如 total_sum）。

In [53]:
M = int( input("請輸入M:") ) # 將提示文字改為 "請輸入M:"
N = int( input("請輸入N:") ) # 將提示文字改為 "請輸入N:"

total_sum = 0 # 將變數名稱改為更具描述性的 total_sum

print("顯示==========================")
print(f"輸入M:{M}")
print(f"輸入N:{N}")

# 迴圈從 1 到 N (包含 N)
for i in range(1, N + 1):
    term = M * i # 計算當前的項 M*i
    total_sum += term # 將 M*i 加到總和中

    if i < N:
        # 如果不是最後一項，後面加上 "+" 號
        print(f"{M}*{i}", end=" + ") # 注意這裡我在 "+" 兩邊加了空格，讓輸出更美觀
    else:
        # 如果是最後一項，後面加上 "=" 號
        print(f"{M}*{i}", end=" = ")

# 迴圈結束後，印出最終的總和
print(total_sum)

請輸入M:5
請輸入N:10
顯示==========================
輸入M:5
輸入N:10
5*1 + 5*2 + 5*3 + 5*4 + 5*5 + 5*6 + 5*7 + 5*8 + 5*9 + 5*10 = 275


In [55]:
#老師的答案
m = int(input('輸入M:'))
n = int(input('輸入N:'))
sum = 0
for num in range(n):
    num += 1
    if num != n:
        print(m,'*',num,' + ',end='')
    else:
        print(m,'*',num,' = ',end='')
        sum += m * num

print(sum)

輸入M:5
輸入N:10
5 * 1  + 5 * 2  + 5 * 3  + 5 * 4  + 5 * 5  + 5 * 6  + 5 * 7  + 5 * 8  + 5 * 9  + 5 * 10  = 50


#字串的處理
##跳脫字元
- \\\' -> 單引號
- \\\" -> 雙引號
- \n -> "換行"
- \t ->  "tab"
- \\\ -> "反斜線"

In [59]:
print("\\")
print("\t嗨")
print("\"")

\
	嗨
"


##原始字串
原始字串的定義
要創建一個原始字串，你只需要在字串的開頭引號前加上一個小寫字母 r (或大寫 R)。

- r'這是個原始字串'
- R"這也是個原始字串"

當你使用 r 前綴時，Python 會告訴解釋器：「嘿，這個字串裡的所有反斜線，都請把它們當作普通的字元，不要再對它們做任何跳脫處理了！」

什麼時候會用到原始字串？
最常見的應用場景是：檔案路徑

In [61]:
>>> print(r'That is Carol\'s cat.')

That is Carol\'s cat.


In [60]:
# 使用三雙引號定義多行字串
multiline_string_double = """Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and
extortion.
Sincerely,
Bob"""

print(multiline_string_double)

print("===============")

# 也可以使用三單引號定義多行字串
multiline_string_single = '''Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and
extortion.
Sincerely,
Bob'''

print(multiline_string_single)

Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and
extortion.
Sincerely,
Bob
Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and
extortion.
Sincerely,
Bob


In [62]:
#單行字串
>>> sentens = "Dear Alice,\
Eve's cat has been arrested for catnapping, cat burglary, and\
 extortion.\
Sincerely,\
Bob"

>>> print(sentens)

Dear Alice,Eve's cat has been arrested for catnapping, cat burglary, and extortion.Sincerely,Bob


In [64]:
#多行字串
>>> sentens = '''Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and extortion.
Sincerely,
Bob'''

>>> print(sentens)


Dear Alice,
Eve's cat has been arrested for catnapping, cat burglary, and extortion.
Sincerely,
Bob


##字串取代
- 語法:字串.replace(old, new[, max])
- replace方法 可以將指定的字串進行更換:
 - 字串.replace(舊字串, 新字串)
 - replace方法 可以指定您要第幾個位置後不進行更換:
- 字串.replace(舊字串, 新字串,替換不可以超過幾次)
 - 位置編號由 0 開始

In [70]:
name = 'Henny'
name.replace('H', 'P')
'P' + name[1:]

'Penny'

##字串內容搜尋
- 語法:字串.find(要搜尋的字串, beg=0, end=len(string))
- 字串可以搜尋特定內容的位置。
- 搜尋的字串位置編號由 0 開始。
- 字串.find(搜尋字串,起始位置,結束位置)
- 搜尋時不包含結束位置。
- 若搜尋不到則回傳 -1。
- 若沒有起始位置，也沒有結束位置，預設由 0 開始。
- 若沒有結束位置，預設到最後一個字。